akd/src/append_only_zks.rs

```diff
@@ -152,15 +153,62 @@ impl Azks {
         storage: &StorageManager<S>,
         insertion_set: &[Node],
     ) -> Result<u64, AkdError> {
-        let prefixes_set = crate::utils::build_prefixes_set(
-            insertion_set
-                .iter()
-                .map(|n| n.label)
-                .collect::<Vec<NodeLabel>>()
-                .as_ref(),
-        );
+        let insertion_labels = insertion_set
+            .iter()
+            .map(|n| n.label)
+            .collect::<Vec<NodeLabel>>();
+
+        self.bfs_preload_nodes_sorted::<S>(storage, &insertion_labels)
+            .await
+    }
 
-        self.bfs_preload_nodes::<S>(storage, prefixes_set).await
+    /// Given a sorted list of [NodeLabel]s, determine if a given [NodeLabel] is a prefix of any in the list.
+    fn insertion_set_contains(insertion_labels: &[NodeLabel], label: &NodeLabel) -> bool {
+        insertion_labels
+            .binary_search_by(|candidate| {
+                match label.label_len == 0 || label.is_prefix_of(candidate) {
+                    true => Ordering::Equal,
+                    false => candidate.label_val.cmp(&label.label_val),
+                }
```

Comment:
Hey @kevinlewi , would we properly handle scenarios like the following?

Say we have this sorted list of labels:

[.., 01, 000, 010, ..]

Say we're looking for 000 and currently at at 01, we would see that 01 is *not* a prefix of 000, then compare the values of 01 and 000. Would we see that 000 is smaller than 01 instead of larger?

Reply:
Ah, insertion sets are labels of the same length, so the above scenario won't happen. That's the assertion here, right?

Reply:
This function assumes that the list of labels passed into it are all of the same length. Indeed, this is the case when using it for the preloading of nodes because the labels are all corresponding to leaf nodes with `label_len = 256`.

So I think the case you mentioned would not occur. Actually, we should add a comment to this function to explicitly clarify this assumption...

Reply:
Ah gotcha. I'm doing some light refactoring to make it clear where inputs need to be sorted (e.g. for `batch_insert_leaves`, many callsites are not sorting the `insertion_set` even though the function assumes it is sorted), I'll take care of adding the comment. Thanks for clarifying!

Reply:
@kevinlewi  Is this assertion true even in auditor directory construction? It seems like the `unchanged_nodes` field in an append only proof contains labels of varying lengths:

https://github.com/novifinancial/akd/blob/50ac0043f1d8b601060310e5768ab560da6d987f/akd/src/auditor.rs

You can pull this branch and run tests, there's one test that fails: https://github.com/afterdusk/akd/tree/test-assertion

Reply:
Ahh, good point, that assumption is not true for the auditor. However, the auditor starts with an empty AZKS and calls batch_insert_leaves, which means that no pre-loading needs to happen (which is why the function isn't broken for auditors).

Hmm... yeah, not entirely sure what to do about this here. It seems like the perf optimization I made might be a little harder to generalize...

Reply:
Ah, that makes sense. Appreciate the clarification! I have an idea for how to encode the sorting + equal length requirements for preloading, I'll take a stab at it and we can discuss on the PR 🙂